[UnifiedPDF] PDF is opening in Safari zoomed in and off-center

smfr opened the pull request:
#### 2c9d242f87f88dd7f4d777ea9f8bc0381345bf03
<pre>
[UnifiedPDF] PDF is opening in Safari zoomed in and off-center
<a href="https://bugs.webkit.org/show_bug.cgi?id=271870">https://bugs.webkit.org/show_bug.cgi?id=271870</a>
<a href="https://rdar.apple.com/125382902">rdar://125382902</a>

Reviewed by Abrar Rahman Protyasha and Sammy Gill.

Safari does per-site saving of page zoom, which means that when you re-open a PDF, it calls `UnifiedPDFPlugin::setPageScaleFactor()`
before we&apos;ve received the PDFDocument. This causes `UnifiedPDFPlugin::setPageScaleFactor()` to set `m_shouldUpdateAutoSizeScale` to `ShouldUpdateAutoSizeScale::No`.

When then run `PDFDocumentLayout::updateLayout()` with a null document, and compute a small maxRowWidth which is just the margins,
thus computing a huge scale. And because `m_shouldUpdateAutoSizeScale` is forever after `ShouldUpdateAutoSizeScale::No`, we get stuck
with this huge scale.

Fix by ensuring that the first layout with a non-null PDFDocument runs with `ShouldUpdateAutoSizeScale::Yes`.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.h:
(WebKit::PDFDocumentLayout::havePDFDocument const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::updateLayout):

Canonical link: <a href="https://commits.webkit.org/276822@main">https://commits.webkit.org/276822@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/698812d22011fda92d8b5bb776a24d107f0f7e5c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45787 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24914 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48368 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48457 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41824 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29198 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22312 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46365 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21995 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39500 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18652 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19400 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3831 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42096 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40946 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50214 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20778 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17287 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/44601 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22083 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43467 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10165 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22442 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21772 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1185,16 +1185,20 @@ static String mutationObserverNotificationString()
     auto layoutSize = availableContentsRect().size();
-    m_documentLayout.updateLayout(layoutSize, m_shouldUpdateAutoSizeScale);
+
+    auto autoSizeMode = m_didLayoutWithValidDocument ? m_shouldUpdateAutoSizeScale : ShouldUpdateAutoSizeScale::Yes;
```
Do you think it would be better to just override the state globally or just keep it local like this?

sammygill replied:
Or maybe not since we would forget about the state change due to the external zoom that occurred

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1185,16 +1185,20 @@ static String mutationObserverNotificationString()
         updateScrollbars();
     }
 
+    m_didLayoutWithValidDocument = m_documentLayout.havePDFDocument();
```
Should we also change the auto size state change underneath this to only set it to `No` if we did a layout with a valid document?

sammygill replied:
(actually probably not due to the same reason above)

smfr committed:
[UnifiedPDF] PDF is opening in Safari zoomed in and off-center
https://bugs.webkit.org/show_bug.cgi?id=271870
rdar://125382902

Reviewed by Abrar Rahman Protyasha and Sammy Gill.

Safari does per-site saving of page zoom, which means that when you re-open a PDF, it calls `UnifiedPDFPlugin::setPageScaleFactor()`
before we've received the PDFDocument. This causes `UnifiedPDFPlugin::setPageScaleFactor()` to set `m_shouldUpdateAutoSizeScale` to `ShouldUpdateAutoSizeScale::No`.

When then run `PDFDocumentLayout::updateLayout()` with a null document, and compute a small maxRowWidth which is just the margins,
thus computing a huge scale. And because `m_shouldUpdateAutoSizeScale` is forever after `ShouldUpdateAutoSizeScale::No`, we get stuck
with this huge scale.

Fix by ensuring that the first layout with a non-null PDFDocument runs with `ShouldUpdateAutoSizeScale::Yes`.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.h:
(WebKit::PDFDocumentLayout::havePDFDocument const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::updateLayout):

Canonical link: https://commits.webkit.org/276822@main
